corehq/apps/geospatial/reports.py

```diff
@@ -116,40 +116,152 @@ class CaseGroupingReport(BaseCaseMapReport):
+        Also determines ``self._precision`` if it is not set.
+        """
+        case_property = get_geo_case_property(self.domain)
 
-        # Apply geohash grid aggregation
-        if 'precision' in self.request.GET:
-            precision = self.request.GET['precision']
+        if self.request.GET.get('precision'):
+            self._precision = int(self.request.GET['precision'])
         else:
-            precision = find_precision(query, case_property)
+            self._precision = find_precision(query, case_property)
 
-        query = apply_geohash_agg(query, case_property, precision)
-        return query
+        return apply_geohash_agg(query, case_property, self._precision)
 
     def _get_geo_point(self, case):
         case_obj = wrap_case_search_hit(case)
         geo_case_property = get_geo_case_property(case_obj.domain)
         geo_point = case_obj.case_json.get(geo_case_property)
         return geo_point
 
+    @property
+    def rows(self):
+        """
+        Returns paginated cases
+        """
+        buckets = self._get_buckets(
+            self.domain,
+            self.shared_pagination_GET_params,
+        )
+        if not buckets:
+            return []
+        # Ignore self.pagination.count. It is always treated as 1.
+        geohash = buckets[self.pagination.start]['key']
+        query_filters = [filters.geo_grid(
+            field=get_geo_case_property(self.domain),
+            geohash=geohash,
+        )]
+        cases = []
+        query = super()._build_query()
+        self._filter_query(query, additional_filters=query_filters)
+        es_results = query.run().raw
+
+        for row in es_results['hits'].get('hits', []):
+            display = CaseDisplayES(
+                self.get_case(row), self.timezone, self.individual
+            )
+            coordinates = self._get_geo_location(self.get_case(row))
+            cases.append([
+                display.case_id,
+                coordinates,
+                display.case_link
+            ])
+        return cases
+
+    @property
+    def total_records(self):
+        """
+        Returns the number of buckets.
+
+        Pagination count is always treated as 1, so the total number of
+        records == the number of pages == the number of buckets.
+        """
+        buckets = self._get_buckets(
+            self.domain,
+            self.shared_pagination_GET_params,
+        )
+        return len(buckets)
+
+    @property
+    def shared_pagination_GET_params(self):
+        shared_params = super().shared_pagination_GET_params
+        shared_params.extend([
+            {'name': 'feature', 'value': self.request.GET.get('feature')},
+            {'name': 'precision', 'value': self._precision},
+        ])
+        return shared_params
+
+    # quickcache uses shared_pagination_GET_params as part of its key
+    # because those determine the results of `query`
+    @quickcache(['self.domain', 'self.shared_pagination_GET_params'],
```

Comment:
:lady_beetle: Bug: `self._precision` is part of the `shared_pagination_GET_params` cache key, but it is set by the `_aggregate_query()` method, which is called by `_get_buckets()`. So the cache key will always be wrong the first time this is called.

Reply:
Resolved: a183cebfb91e70f208c4f72b1083db2e9ae04b2f